### ISSUE TITLE
`while` loop "not allowed outside of blocks"

### ISSUE THREAD

**Issue by Braqzen** [FuelLabs/sway#2455]
Is this a bug?
What is a "block" here?

```rust
contract;

abi MyContract {
    fn test_function() -> bool;
}

impl MyContract for Contract {
    fn test_function() -> bool {
        let mut counter = 0;
        let truth = false;
        let result = while counter < 10 {
            counter += 1;
            true
        };
        result
    }
}
```

![1](https://user-images.githubusercontent.com/103777923/182673464-06293591-ec90-40e6-84e7-5a95d2289e77.png)

**Comment by mohammadfawaz** [comment]
Yeah don't get that either. Marking this as critical. Either this should work or it should emit a much better error message. And I don't see why it shouldn't work...

**Comment by otrho** [comment]
Are `while` loops expressions?  I was under the impression their bodies couldn't return a type.  :thinking:  If they turn out to be statements then only allowing them as direct children of code blocks makes sense, but that shouldn't be the error message.

**Comment by otrho** [comment]
Personally I don't like the idea of having statements as a special case, but having an expression which must return `()` is the same, but better because it can type-check.  So I think what Rust does is right; the body must have the type `()` and so does the `while` loop.
